### PR TITLE
Adding Lab: Continuous Training with TensorFlow, PyTorch, XGBoost, and Scikit Learn Models with Kubeflow and AI Platform Pipelines

### PR DESCRIPTION
**Pull request by kylesteckler**
Adding lab and solution (Continuous Training with TensorFlow, PyTorch, XGBoost, and Scikit Learn Models with Kubeflow and AI Platform Pipelines) for MLOps Course 2

### CODE REVIEW

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 1]
```diff
@@ -0,0 +1,1075 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Lab: Continuous Training with TensorFlow, PyTorch, XGBoost, and Scikit Learn Models with KubeFlow and AI Platform Pipelines\n",
+    "\n",
+    "In this lab we will create containerized training applications for ML models in TensorFlow, PyTorch, XGBoost, and Scikit Learn. Will will then use these images as ops in a KubeFlow pipeline and train multiple models in parallel. We will then set up recurring runs of our KubeFlow pipeline in the UI. \n",
+    "\n",
+    "First, we will containerize models in TF, PyTorch, XGBoost and Scikit Learn following a step-wise process for each:\n",
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:11]:
```suggestion
    "First, we will containerize models in TF, PyTorch, XGBoost and Scikit-learn following a step-wise process for each:\n",
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 2]
```diff
@@ -0,0 +1,1075 @@
...
+    "AND functional_weight IS NOT NULL"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create Scikit Training Script\n",
+    "\n",
+    "We will develop our first training script with Scikit Learn. We will use Pandas to read the CSV data then train a simple [SGD Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html)."
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:87]:
```suggestion
    "We will develop our first training script with Scikit-learn. We will use Pandas to read the CSV data then train a simple [SGD Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html)."
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 3]
```diff
@@ -0,0 +1,1075 @@
...
+    "AND hours_per_week IS NOT NULL\n",
+    "AND income_bracket IS NOT NULL \n",
+    "AND functional_weight IS NOT NULL"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create Scikit Training Script\n",
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:85]:
```suggestion
    "### Create Scikit-learn Training Script\n",
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 4]
```diff
@@ -0,0 +1,1075 @@
...
+    "\n",
+    "if __name__ == '__main__':\n",
+    "    fire.Fire(train_evaluate)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Package SciKit Training Script into a Docker Image\n",
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:171]:
```suggestion
    "### Package Scikit-learn training script into a Docker image\n",
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,1075 @@
...
+    "The next step is to package this training script into a Docker Image. We need to be sure to list the dependencies in this Dockerfile. For the Scikit model we need to ensure scikit-learn version 0.23.2 and Pandas version 1.1.1"
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:172]:
```suggestion
    "The next step is to package this training script into a Docker image. We need to be sure to list the dependencies in this Dockerfile. For the Scikit-learn model we need to ensure scikit-learn version 0.23.2 and Pandas version 1.1.1"
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,1075 @@
...
+    "COPY train.py .\n",
+    "\n",
+    "ENTRYPOINT [\"python\", \"train.py\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Build the Scikit Learn Trainer Image\n",
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:195]:
```suggestion
    "### Build the Scikit-learn trainer image\n",
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 7]
```diff
@@ -0,0 +1,1075 @@
...
+    "if __name__ == '__main__':\n",
+    "    fire.Fire(train_evaluate)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Package TensorFlow Training Script into a Docker Image\n",
+    "Note that the dependencies in this Dockerfile are different than for the Scikit learn one. "
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:366]:
```suggestion
    "Note that the dependencies in this Dockerfile are different than for the Scikit-learn one. "
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 8]
```diff
@@ -0,0 +1,1075 @@
...
+   "metadata": {},
+   "source": [
+    "## Develop KubeFlow Pipeline\n",
+    "Now that you have all four of your training applications as containers in your project's Container Registry, let's build a KubeFlow pipeline. \n",
+    "\n",
+    "The KubeFlow pipeline will have two BigQuery Ops. We will use the pre-built BigQuery Query component (no need to reinvent the wheel) to do the following: \n",
+    "* Create a training split in our data and export to CSV\n",
+    "* Create a validation split in our data and export to CSV\n",
+    "\n",
+    "The output of these BigQuery Ops will be the input data into four AI Platform Training Ops. For this we will also use a pre-built component. Each AI Platform Training Op will train one of our containerized models - Tensorflow, PyTorch, XGBoost, and Scikit Learn."
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:753]:
```suggestion
    "The output of these BigQuery Ops will be the input data into four AI Platform Training Ops. For this we will also use a pre-built component. Each AI Platform Training Op will train one of our containerized models - Tensorflow, PyTorch, XGBoost, and Scikit-learn."
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 9]
```diff
@@ -0,0 +1,1075 @@
...
+    "    tf_args = [\n",
+    "        '--training_dataset_path', create_training_split.outputs['output_gcs_path'],\n",
+    "        '--validation_dataset_path', create_validation_split.outputs['output_gcs_path'],\n",
+    "        '--output_dir', tf_output_dir,\n",
+    "        '--batch_size', '32', \n",
+    "        '--num_train_examples', '1000',\n",
+    "        '--num_evals', '10'\n",
+    "    ]\n",
+    "    \n",
+    "    # Training arguments to be passed to the Scikit Trainer\n",
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:860]:
```suggestion
    "    # Training arguments to be passed to the Scikit-learn Trainer\n",
```

File: continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb [hunk 10]
```diff
@@ -0,0 +1,1075 @@
...
+   "source": [
+    "print(f\"https://{ENDPOINT}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### End of Lab, Congrats!\n",
+    "In this lab you created containerized training applications for TensorFlow, PyTorch, Scikit Learn, an XGBoost models. You then created a KubeFlow pipeline that used pre built components to create training/validation splits in BigQuery data, export that data as CSV files to GCS, and launch AI Platform Training Jobs with the four containerized training applications. Finally, you ran the pipeline from the UI and set up Continuous Training to re-run your pipeline once a week!"
```
Comment (dougkelly) [inline continuous_training/kubeflow/solutions/multiple_frameworks_kubeflow.ipynb:1045]:
```suggestion
    "In this lab you created containerized training applications for TensorFlow, PyTorch, Scikit-learn, an XGBoost models. You then created a KubeFlow pipeline that used pre built components to create training/validation splits in BigQuery data, export that data as CSV files to GCS, and launch AI Platform Training Jobs with the four containerized training applications. Finally, you ran the pipeline from the UI and set up Continuous Training to re-run your pipeline once a week!"
```

**Review by dougkelly — APPROVED** [review]
Hey Kyle,

Great job, really great lab for showing the cross-framework flexibility of Kubeflow / AI Platform Pipelines!

I received FileNotFoundError: [Errno 2] No such file or directory: './scikit_trainer_image/train.py' errors when running through the %%writefile to new file path magics as FYI. Is it possible that the directory needs to be created first before the file is written there?

A few small nits; I added suggested fixes to keep the Scikit-learn formatting and casing consistent throughout.

Approving. Feel free to address any comments before merging or after on subsequent iterations.

**Comment by kylesteckler** [comment]
@googlebot I fixed it.

**Pull request merged**